### PR TITLE
[camera] Allow logical cameras to use all physical cameras via zoom on android 11+

### PR DESCRIPTION
**Pull request by lucasoskorep**
Cameras in android versions greater than 11 are strongly pushed by the android team to use logical cameras to wrap multiple physical rear cameras into a single "logical camera."  

The Flutter camera plugin has an issue where it calculates zoom incorrectly on these cameras and stops the user from being able to change the zoom appropriately to swap between the physical cameras wrapped by the logical camera exposed via the Android API

This change uses the new APIs implemented in Android 11 to correctly calculate zoom, and allow the camera app to swap between all the physical cameras contained within a single Logical camera in android. Users will still see a single Rear cam, but zooming should swap correctly. 

Linked to https://github.com/flutter/plugins/pull/6091

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android/CHANGELOG.md
Comment (camsim99) [inline packages/camera/camera_android/CHANGELOG.md:3]: Nit, address zoom issue and capitalize Android, something like:

```suggestion
* Fixes zoom computation for virtual cameras hiding physical cameras in Android 11+.
```

File: packages/camera/camera_android/android/build.gradle
Comment (camsim99) [inline packages/camera/camera_android/android/build.gradle:62]: Is this needed?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java:150]: Why remove the periods here and below? I think everything should be made consistent and looks like most things end in periods?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java:41]: Nit: can this refer to a constant defined above so it's not a magic number?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java:39]: Can you leave a comment somewhere around here as to why this is done for API 30+?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomUtils.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomUtils.java:31]: Nit: can this refer to the helper method you wrote below since they are the same?

File: packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeatureTest.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeatureTest.java:46]: Do you think we can add a test here that utilizes Robolectric to make sure the proper `CaptureRequest` is sent when `updateBuilder` is called?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraZoom.java
Comment (camsim99) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraZoom.java:1]: I agree that this and its corresponding test should be deleted, since this doesn't seem to be used anywhere. @stuartmorgan can you double check me on this? I'm not sure if you have more context.

**Review by camsim99 — CHANGES_REQUESTED** [review]
Thank you for doing this! I see this method you used is preferred for Android 11+ so this is great! I just requested some changes and had some questions.

File: packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeatureTest.java
Comment (lucasoskorep) [reply packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeatureTest.java:49]: So Robolectric has some major issues mocking this call 

`mock(CaptureRequest.Builder.class);`

Seems it can only be mocked using our standard test runner - checking the rest of the project nothing mocks this while using mockito. 

I did however create a small method to help mock which version of the SDK_INT is running in our code and was able to write a few more tests using this to set our build SDK to R or Q depending on the test.

Comment (camsim99) [reply packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeatureTest.java:49]: Oh gotcha. Okay thanks!

File: packages/camera/camera_android/CHANGELOG.md
Comment (camsim99) [inline packages/camera/camera_android/CHANGELOG.md:1]: Nit: maybe mention deleting `CameraZoom` here, as well?

**Review by camsim99 — APPROVED** [review]
Left a couple more comments, but looks good overall! @stuartmorgan or @bparrishMines pinging you for second review!

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java
Comment (lucasoskorep) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java]: 👍 Added

**Comment by benkuper** [comment]
+1 ! This is a high priority necessity for me and it seems that the PR could go on ?
If not yet, how can I test this before the PR is merged ? Sorry, not a Flutter veteran so trying to understand how to use the modified package to test the functionality.
Thank you !

**Comment by lucasoskorep** [comment]
> +1 ! This is a high priority necessity for me and it seems that the PR could go on ? If not yet, how can I test this before the PR is merged ? Sorry, not a Flutter veteran so trying to understand how to use the modified package to test the functionality. Thank you !
> 
> EDIT : using the following seems to allow me to test the modified plugin, but not sure where to go from there, is there anything to add to the code or is the zoom detection / lens switch automatic ?
> 
> ```
> dependency_overrides:
>   camera_android:
>     git:
>       dependency: transitive
>       url: https://github.com/lucasoskorep/plugins.git
>       ref: feat/allowCameraZoomToUseAllPhysicalCameras
>       path: packages/camera/camera_android/
> 
>   camera:
>     git:
>       url: https://github.com/lucasoskorep/plugins.git
>       ref: feat/allowCameraZoomToUseAllPhysicalCameras
>       path: packages/camera/camera/ 
> ```

This is the fix I have for my application for now.  Still pending @stuartmorgan's final approval though so sit tight for getting it from the main package itself! 

As for functionality any android phone with a logical camera hiding multiple physical cameras will just return the correct min and max zoom values with the fix so no code changes outside of potentially how you are guarding min and max zoom values :)

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java
Comment (stuartmorgan) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java:158]: Nit: missing period here and in the `@return` section; same in the other new method.

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraZoom.java
Comment (stuartmorgan) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraZoom.java:1]: Looks like it was replaced with `ZoomLevelFeature` during a refactor and the dead code just accidentally wasn't cleaned up.

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java
Comment (stuartmorgan) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java:163]: I assume this should say "minimum" rather than "maxmium"?

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java
Comment (stuartmorgan) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/features/zoomlevel/ZoomLevelFeature.java:15]: The previous naming style was correct per https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Just a few minor nits, but otherwise looks good! (With the caveat that I'm not very familiar with the camera APIs, so I can't do a in-depth review on the substance of the change.)

**Comment by lucasoskorep** [comment]
@stuartmorgan thanks for reviewing the PR!  Got the requested changes added in and updated the version to +3 in pubspec and changelog.md so this PR would be up to date.  Let me know if there is anything else you'd like added and I can get fixes set up for it :)

**Comment by GaryQian** [comment]
@stuartmorgan This seems to be ready for another look

**Review by stuartmorgan — APPROVED** [review]
Sorry for the really long delay getting back to this; I lost track of the review request :(

**Pull request merged**
